drivers/scheduler/openshift/openshift.go

```diff
@@ -137,6 +148,146 @@ func (k *openshift) updateSecurityContextConstraints(namespace string) error {
 }
 
+func (k *openshift) UpgradeScheduler(version string) error {
+	var output []byte
+	var err error
+
+	if err = downloadOCP4Client(version); err != nil {
+		return err
+	}
+	if output, err = exec.Command("oc", "version").CombinedOutput(); err != nil {
+		return err
+	}
+	logrus.Infof("output: %s", output)
+
+	channel := ""
+	if channel, err = getChannel(version); err != nil {
+		return err
+	}
+	logrus.Infof("Selected channel: %s", channel)
+
+	patch := `
+spec:
+  channel: %s
+`
+	args := []string{"patch", "clusterversion", "version", "--type=merge", "--patch", fmt.Sprintf(patch, channel)}
+	if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
+		return err
+	}
+	logrus.Infof("output: %s", output)
+
+	upgradeArg := fmt.Sprintf("--to=%s", version)
+
+	if versionReg.MatchString(version) {
+		upgradeArg = "--to-latest=true"
+	}
+
+	args = []string{"adm", "upgrade", upgradeArg}
+	if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
+		return err
+	}
+	logrus.Infof("Upgrade started: %s", output)
+
+	args = []string{"get", "clusterversion", "version", "-o=jsonpath='{.spec.desiredUpdate.version}'"}
+
+	if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
+		return err
+	}
+
+	desiredVersion := strings.TrimSpace(string(output))
+	desiredVersion = strings.Trim(desiredVersion, "\"")
+	desiredVersion = strings.Trim(desiredVersion, "'")
+
+	t := func() (interface{}, bool, error) {
+		jsonpath := fmt.Sprintf("{.status.history[?(@.version==\"%s\")].state}", desiredVersion)
+		args := []string{"get", "clusterversion", "version", fmt.Sprintf("-o=jsonpath='%s'", jsonpath)}
+		if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
+			return nil, true, fmt.Errorf("failed to get cluster version. cause: %v", err)
+		}
+		state := strings.TrimSpace(string(output))
+		state = strings.Trim(state, "\"")
+		state = strings.Trim(state, "'")
+		if state != "Completed" {
+			jsonpath = "{.status.conditions[?(@.type==\"Progressing\")].message}"
+			args := []string{"get", "clusterversion", "version", fmt.Sprintf("-o=jsonpath='%s'", jsonpath)}
+			if output, err = exec.Command("oc", args...).CombinedOutput(); err != nil {
+				return nil, true, fmt.Errorf("failed to get cluster condition. cause: %v", err)
+			}
+			return nil, true, fmt.Errorf("cluster not upgraded yet. cause: %s", string(output))
+		}
+		return nil, false, nil
+	}
+
+	if _, err := task.DoRetryWithTimeout(t, 2*time.Hour, 15*time.Second); err != nil {
+		return err
+	}
+	logrus.Infof("Cluster is now %s", desiredVersion)
+	return nil
+}
+
+func getChannel(version string) (string, error) {
+	if versionReg.MatchString(version) {
+		return version, nil
+	}
+
+	ver, err := semver.Make(version)
+	if err != nil {
+		return "", err
+	}
+	channel := fmt.Sprintf("candidate-%d.%d", ver.Major, ver.Minor)
+	url := fmt.Sprintf("curl -sL https://mirror.openshift.com/pub/openshift-v4/clients/ocp/%s | grep %s", channel, version)
```

Comment:
We can define this as a constant:
```
openshiftmirror="https://mirror.openshift.com/pub/openshift-v4/clients/ocp"
```
Will make the code more readable.

Reply:
done